### ISSUE TITLE
Incorrect route generated for page name with underscore

### ISSUE THREAD

**Issue by ianpurvis** [nuxt/nuxt.js#2716]
Hi, I found an issue with route generation for pages that contain an underscore in the name.

When page name is `pages/sept_2017.vue`
Observed route path is `/sept:2017?`
Expected route path to be `/sept_2017`

I'm testing by inspecting `this.$nuxt._router.options.routes` in the browser:

```
{path: "/sept:2017?", component: function, name: "sept2017"}
```

The main problem is that the `?` wildcard matches invalid paths like `sept_2017njakljsndflkajsndflkajsdf`, which breaks 404 routing.

You can work around the problem by removing the underscore from the page filename.

**Comment by Zip753** [comment]
First of all, I think it's quite uncommon to use underscore in URLs, so you might want to rethink that. The issue is that Nuxt.js treats underscores as start of the name of router param, hence the path you observe.

Now, you can either use custom routing or try to override the route yourself in `router.extendRoutes` config, check out the docs for usage: https://nuxtjs.org/api/configuration-router/#extendroutes. While the easiest way to resolve that would be... well, to change underscore to hyphen :) (it's [better for SEO!](https://stackoverflow.com/q/119312))

**Comment by ianpurvis** [comment]
@Zip753 Yeah, fair point... I might move towards hyphens regardless. I have been using custom routing to make it work. However, the behavior still feels a little wrong to me because I thought the dynamic routing would be driven by leading underscore only, per https://nuxtjs.org/guide/routing#dynamic-routes

> To define a dynamic route with a param, you need to define a .vue file OR a directory **prefixed by an underscore**.

So, if the current behavior is ok, maybe the doc could help clarify.

Also, I am happy to try a patch to the route generation code, if someone can point me in the right direction... like here? https://github.com/nuxt/nuxt.js/blob/dev/lib/common/utils.js#L257

**Comment by David-Mueller** [comment]
@Zip753 
Is it a feature, that files or folders containing underscores in the middle of the name lead to route params being extracted? (As the doc states `prefixed by an underscore`.)

Currently, folder structures like `pages/test_route` lead to `this.$route.params = {route: "_route"}` for requests to `http://server/test_route`.
If this is a bug, I'd give it a shot in fixing it as well.

Cheers

**Comment by David-Mueller** [comment]
Hey there,

not an easy call. It might be a backward compatibility problem to just remove the feature. (See the MR)
I guess it's a decision between removing the feature or updating the documentation.

I've never seen a router supporting those "inline route params" as far as I remember. Maybe it's enough to support it via extendRoutes and allow snake case page names by default.

**Comment by fzembow** [comment]
FWIW I too found this confusing, particularly the behavior when you have multiple routes with an underscore routing to the first alphabetically. In my situation, the files looked like:

`pages/writing/the_first_thing.vue`
`pages/writing/the_second_thing.vue`

that would mysteriously display `the_first_thing` when navigating to `/writing/the_second_thing` ... very surprising behavior

My vote would be to only do dynamic mapping when prefixed, I would suppose that dynamic parameters in the middle of a route would be exceedingly rare.

While it seems there's some consensus that hyphens are "better" for SEO, I don't see this as some kind of widely acknowledged web standard that makes sense to build into a framework as flexible and awesome as nuxt :)

**Comment by aldarund** [comment]
@galvez @Atinux it seems to be solved in current edge
```pages/auth/test_me_new.vue```
Generated route
```
                {
			path: "/auth/test_me_new",
			component: _a2e9bbd6,
			name: "auth-test_me_new"
		},
```
And all examples from this topic work fine for me
